source/material_model/reactive_fluid_transport.cc

```diff
@@ -181,15 +186,18 @@ namespace aspect
               && in.requests_property(MaterialProperties::reaction_rates))
             {
               std::vector<double> eq_free_fluid_fractions(out.n_evaluation_points());
-              melt_fractions(in, eq_free_fluid_fractions);
+              melt_fractions(in, eq_free_fluid_fractions, out);
 
               for (unsigned int q=0; q<out.n_evaluation_points(); ++q)
                 for (unsigned int c=0; c<in.composition[q].size(); ++c)
                   {
-                    double porosity_change = eq_free_fluid_fractions[q] - in.composition[q][porosity_idx];
+                    const double volume_frac_porosity = in.composition[q][porosity_idx];
+                    const double bulk_density = out.densities[q] * (1.0 - volume_frac_porosity) + fluid_out->fluid_densities[q] * volume_frac_porosity;
+                    const double mass_frac_porosity = volume_frac_porosity * fluid_out->fluid_densities[q] / bulk_density;
+                    double porosity_change = eq_free_fluid_fractions[q] - mass_frac_porosity;
                     // do not allow negative porosity
-                    if (in.composition[q][porosity_idx] + porosity_change < 0)
-                      porosity_change = -in.composition[q][porosity_idx];
+                    if (mass_frac_porosity + porosity_change < 0)
+                      porosity_change = -mass_frac_porosity;
```

Comment:
Don't you have to convert it back to volume fraction before using it in the reaction rates below?